### PR TITLE
SNOW-911146  Extract method of preparing connection in tests

### PR DESCRIPTION
**Pull request by sfc-gh-pmotacki**
### Description
Please explain the changes you made here.

### Checklist
- [x] Code compiles correctly
- [x] Run ``make fmt`` to fix inconsistent formats
- [x] Run ``make lint`` to get lint errors and fix all of them
- [ ] Created tests which fail without the change (if possible)
- [x] All tests passing
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: connection_test.go [hunk 1]
```diff
@@ -398,70 +398,49 @@ func TestPrivateLink(t *testing.T) {
...
-		t.Error(err)
-	}
+		table(generator(rowcount => 100)))`, nil)
 
-	rows, err := sc.QueryContext(ctx, "select min(c1) as ms, sum(c1) from ut_conn group by (c1 % 10) order by ms", nil)
-	if err != nil {
-		t.Error(err)
-	}
-	qid := rows.(SnowflakeResult).GetQueryID()
+		rows, err := sct.sc.QueryContext(ctx, "select min(c1) as ms, sum(c1) from ut_conn group by (c1 % 10) order by ms", nil)
```
Comment (sfc-gh-pfus) [inline connection_test.go:408]: Maybe we can introduce `mustQueryContext` here?

File: connection_test.go [hunk 2]
```diff
@@ -575,156 +554,107 @@ func executeQueryAndConfirmMessage(db *sql.DB, query string, expectedErrorTable
...
-		FuncPostQuery:  postQueryTest,
-		TokenAccessor:  getSimpleTokenAccessor(),
-		RequestTimeout: 10,
-	}
-
-	sc.rest = sr
+	runSnowflakeConnTest(t, func(sct *SCTest) {
+		sc := sct.sc
+		ctx := sc.ctx
+		sr := &snowflakeRestful{
```
Comment (sfc-gh-pfus) [inline connection_test.go:589]: It works without closing func?

File: connection_test.go [hunk 3]
```diff
@@ -575,156 +554,107 @@ func executeQueryAndConfirmMessage(db *sql.DB, query string, expectedErrorTable
...
+			TokenAccessor:  getSimpleTokenAccessor(),
+			RequestTimeout: 10,
+		}
+		sc.rest = sr
 
-	_, err = sc.ExecContext(ctx, "SELECT 1", []driver.NamedValue{})
-	if err == nil {
-		t.Fatalf("should have raised an error")
-	}
+		_, err := sc.ExecContext(ctx, "SELECT 1", []driver.NamedValue{})
```
Comment (sfc-gh-pfus) [inline connection_test.go:596]: Maybe we can introduce `mustExecContext` here?

File: connection_test.go [hunk 4]
```diff
@@ -575,156 +554,107 @@ func executeQueryAndConfirmMessage(db *sql.DB, query string, expectedErrorTable
...
-	if tx == nil {
-		t.Fatal("should have created a transaction with connection")
-	}
-	sc.Close()
+	runSnowflakeConnTest(t, func(sct *SCTest) {
+		tx, _ := sct.sc.Begin()
+		if tx == nil {
+			t.Fatal("should have created a transaction with connection")
+		}
+		sct.sc.Close()
```
Comment (sfc-gh-pfus) [inline connection_test.go:658]: We don't need it, `runSnowflakeConnTest` will close sc.

File: connection_test.go (hunk 2 above)
Comment (sfc-gh-pmotacki) [reply connection_test.go:589]: True This wrapping function requires closing function to close connection at the end of execution

File: connection_test.go (hunk 1 above)
Comment (sfc-gh-pmotacki) [reply connection_test.go:408]: Not in this case. We need error here.

File: connection_test.go (hunk 3 above)
Comment (sfc-gh-pmotacki) [reply connection_test.go:597]: Not in this case. We need error here.

File: connection_test.go (hunk 4 above)
Comment (sfc-gh-pmotacki) [reply connection_test.go]: Ok

File: connection_test.go [hunk 5]
```diff
@@ -398,70 +398,46 @@ func TestPrivateLink(t *testing.T) {
...
-	}
-	if err = authenticateWithConfig(sc); err != nil {
-		t.Error(err)
-	}
+	runSnowflakeConnTest(t, func(sct *SCTest) {
+		ctx := sct.sc.ctx
 
-	if _, err = sc.Exec(`create or replace table ut_conn(c1 number, c2 string)
+		sct.mustExec(`create or replace table ut_conn(c1 number, c2 string)
 		as (select seq4() as seq, concat('str',to_varchar(seq)) as str1 from
```
Comment (sfc-gh-pfus) [inline connection_test.go:405]: Can we preserve some indentation in this string?

File: connection_test.go [hunk 6]
```diff
@@ -398,70 +398,46 @@ func TestPrivateLink(t *testing.T) {
...
-	if err != nil {
-		t.Error(err)
-	}
-	if err = authenticateWithConfig(sc); err != nil {
-		t.Error(err)
-	}
-
-	sc.rest.RequestTimeout = 1 * time.Second
+	runSnowflakeConnTest(t, func(sct *SCTest) {
+		ctx := sct.sc.ctx
```
Comment (sfc-gh-pfus) [inline connection_test.go:433]: Inline?

File: connection_test.go [hunk 7]
```diff
@@ -575,156 +551,102 @@ func executeQueryAndConfirmMessage(db *sql.DB, query string, expectedErrorTable
...
-		TokenAccessor:  getSimpleTokenAccessor(),
-		RequestTimeout: 10,
-	}
-	sc.rest = sr
-	_, err = sc.QueryArrowStream(ctx, query)
-	if err == nil {
-		t.Error("should have raised an error")
-	}
+	runSnowflakeConnTest(t, func(sct *SCTest) {
+		sc := sct.sc
```
Comment (sfc-gh-pfus) [inline connection_test.go:555]: Inline?

File: connection_test.go [hunk 8]
```diff
@@ -575,156 +551,102 @@ func executeQueryAndConfirmMessage(db *sql.DB, query string, expectedErrorTable
...
-
-	sr := &snowflakeRestful{
-		FuncPostQuery:  postQueryTest,
-		TokenAccessor:  getSimpleTokenAccessor(),
-		RequestTimeout: 10,
-	}
-
-	sc.rest = sr
+	runSnowflakeConnTest(t, func(sct *SCTest) {
+		sc := sct.sc
```
Comment (sfc-gh-pfus) [inline connection_test.go:584]: Inline both vars?

File: connection_test.go [hunk 9]
```diff
@@ -575,156 +551,102 @@ func executeQueryAndConfirmMessage(db *sql.DB, query string, expectedErrorTable
...
+		ctx := sc.ctx
+		sr := &snowflakeRestful{
```
Comment (sfc-gh-pfus) [inline connection_test.go:586]: Inline?

File: connection_test.go [hunk 10]
```diff
@@ -575,156 +551,102 @@ func executeQueryAndConfirmMessage(db *sql.DB, query string, expectedErrorTable
...
-
-	sr := &snowflakeRestful{
-		FuncPostQuery:  postQueryTest,
-		TokenAccessor:  getSimpleTokenAccessor(),
-		RequestTimeout: 10,
-	}
-
-	sc.rest = sr
+	runSnowflakeConnTest(t, func(sct *SCTest) {
+		ctx := sct.sc.ctx
```
Comment (sfc-gh-pfus) [inline connection_test.go:609]: Inline?

File: connection_test.go [hunk 11]
```diff
@@ -575,156 +551,102 @@ func executeQueryAndConfirmMessage(db *sql.DB, query string, expectedErrorTable
...
+			FuncCloseSession: closeSessionMock,
+			TokenAccessor:    getSimpleTokenAccessor(),
+			RequestTimeout:   10,
+		}
 
-	_, err = sc.QueryContext(ctx, "SELECT 1", []driver.NamedValue{})
-	if err == nil {
-		t.Fatalf("should have raised an error")
-	}
+		sct.sc.rest = sr
```
Comment (sfc-gh-pfus) [inline connection_test.go:617]: Inline?

File: multistatement_test.go [hunk 12]
```diff
@@ -481,115 +481,97 @@ func funcGetQueryRespError(_ context.Context, _ *snowflakeRestful, _ *url.URL, _
...
-	data = &execResponseData{
-		ResultIDs:   "1eFhmhe23242kmfd540GgGre,1eFhmhe23242kmfd540GgGre",
-		ResultTypes: "12544,12544",
-	}
-	sc.rest = sr
-	_, err = sc.handleMultiExec(context.Background(), *data)
-	if err == nil {
-		t.Fatalf("should have failed")
-	}
+		sr := &snowflakeRestful{
```
Comment (sfc-gh-pfus) [inline multistatement_test.go:501]: Inline?

File: multistatement_test.go [hunk 13]
```diff
@@ -481,115 +481,97 @@ func funcGetQueryRespError(_ context.Context, _ *snowflakeRestful, _ *url.URL, _
...
+			FuncGet:       funcGetQueryRespFail,
+			TokenAccessor: getSimpleTokenAccessor(),
+		}
+		data = &execResponseData{
+			ResultIDs:   "1eFhmhe23242kmfd540GgGre,1eFhmhe23242kmfd540GgGre",
+			ResultTypes: "12544,12544",
+		}
+		sct.sc.rest = sr
+		_, err = sct.sc.handleMultiExec(context.Background(), *data)
```
Comment (sfc-gh-pfus) [inline multistatement_test.go:510]: Maybe we can skip `*` here and `&` on init?

File: multistatement_test.go [hunk 14]
```diff
@@ -481,115 +481,97 @@ func funcGetQueryRespError(_ context.Context, _ *snowflakeRestful, _ *url.URL, _
...
+			t.Fatalf("should have failed")
+		}
+		driverErr, ok := err.(*SnowflakeError)
+		if !ok {
+			t.Fatalf("should be snowflake error. err: %v", err)
+		}
+		if driverErr.Number != ErrNoResultIDs {
+			t.Fatalf("unexpected error code. expected: %v, got: %v", ErrNoResultIDs, driverErr.Number)
+		}
+		sr := &snowflakeRestful{
```
Comment (sfc-gh-pfus) [inline multistatement_test.go:549]: Inline?

File: multistatement_test.go [hunk 15]
```diff
@@ -481,115 +481,97 @@ func funcGetQueryRespError(_ context.Context, _ *snowflakeRestful, _ *url.URL, _
...
+			FuncGet:       funcGetQueryRespFail,
+			TokenAccessor: getSimpleTokenAccessor(),
+		}
+		data = &execResponseData{
```
Comment (sfc-gh-pfus) [inline multistatement_test.go:553]: Same here about `&` and `*`?

File: put_get_with_aws_test.go [hunk 16]
```diff
@@ -218,50 +206,41 @@ func TestPretendToPutButList(t *testing.T) {
...
-	if err != nil {
-		t.Fatal(err)
-	}
-	if err = authenticateWithConfig(sc); err != nil {
-		t.Fatal(err)
-	}
+	runSnowflakeConnTest(t, func(sct *SCTest) {
+		sc := sct.sc
+		tableName := randomString(5)
+		if _, err = sc.Exec("create or replace table "+tableName+
```
Comment (sfc-gh-pfus) [inline put_get_with_aws_test.go:212]: `mustExec`?

File: rows_test.go [hunk 17]
```diff
@@ -452,40 +452,28 @@ func TestDownloadChunkErrorStatus(t *testing.T) {
...
-	}
-	defer sc.Exec("drop table if exists testArrowBatches", nil)
+		sct.mustExec("create or replace table testArrowBatches (a int, b int)", nil)
+		defer sct.sc.Exec("drop table if exists testArrowBatches", nil)
 
-	result, err := sc.ExecContext(ctx, "insert into testArrowBatches values (1, 2), (3, 4), (5, 6)", []driver.NamedValue{})
-	if err != nil {
-		t.Error(err)
-	}
+		result, err := sct.sc.ExecContext(ctx, "insert into testArrowBatches values (1, 2), (3, 4), (5, 6)", []driver.NamedValue{})
```
Comment (sfc-gh-pfus) [inline rows_test.go:460]: `mustExecContext`?

File: telemetry_test.go [hunk 18]
```diff
@@ -14,332 +14,255 @@ import (
...
-		t.Errorf("there should be 1 telemetry data in log. found: %v", len(st.logs))
-	}
-	if sendErr := st.sendBatch(); sendErr != nil {
-		t.Fatal(sendErr)
-	}
-	if len(st.logs) != 0 {
-		t.Errorf("there should be no telemetry data in log. found: %v", len(st.logs))
-	}
+	runSnowflakeConnTest(t, func(sct *SCTest) {
+		st := &snowflakeTelemetry{
```
Comment (sfc-gh-pfus) [inline telemetry_test.go:49]: Inline?

File: telemetry_test.go [hunk 19]
```diff
@@ -14,332 +14,255 @@ import (
...
-		TokenAccessor: getSimpleTokenAccessor(),
-	}
-	st := &snowflakeTelemetry{
-		sr:        sr,
-		mutex:     &sync.Mutex{},
-		enabled:   true,
-		flushSize: defaultFlushSize,
-	}
+	runSnowflakeConnTest(t, func(sct *SCTest) {
+		sr := &snowflakeRestful{
```
Comment (sfc-gh-pfus) [inline telemetry_test.go:107]: Inline?

File: telemetry_test.go [hunk 20]
```diff
@@ -14,332 +14,255 @@ import (
...
```
Comment (sfc-gh-pfus) [inline telemetry_test.go:137]: Inline?

File: telemetry_test.go [hunk 21]
```diff
@@ -14,332 +14,255 @@ import (
...
-	}); err != nil {
-		t.Fatal(err)
-	}
-	st.enabled = false
-	err = st.sendBatch()
-	if err == nil {
-		t.Fatal("should have failed")
-	}
+	runSnowflakeConnTest(t, func(sct *SCTest) {
+		sr := &snowflakeRestful{
```
Comment (sfc-gh-pfus) [inline telemetry_test.go:207]: Inline?

File: telemetry_test.go [hunk 22]
```diff
@@ -14,332 +14,255 @@ import (
...
-	if err = authenticateWithConfig(sc); err != nil {
-		t.Fatal(err)
-	}
-
-	sr := &snowflakeRestful{
-		FuncPost:      funcPostTelemetryRespFail,
-		TokenAccessor: getSimpleTokenAccessor(),
-	}
+	runSnowflakeConnTest(t, func(sct *SCTest) {
+		sr := &snowflakeRestful{
```
Comment (sfc-gh-pfus) [inline telemetry_test.go:246]: Inline?

**Review by sfc-gh-pfus — APPROVED** [review]

File: connection_test.go (hunk 11 above)
Comment (sfc-gh-pmotacki) [reply connection_test.go:617]: ok

File: multistatement_test.go (hunk 12 above)
Comment (sfc-gh-pmotacki) [reply multistatement_test.go:501]: ok

File: multistatement_test.go (hunk 13 above)
Comment (sfc-gh-pmotacki) [reply multistatement_test.go:510]: ok

File: multistatement_test.go (hunk 14 above)
Comment (sfc-gh-pmotacki) [reply multistatement_test.go:549]: ok

File: multistatement_test.go (hunk 15 above)
Comment (sfc-gh-pmotacki) [reply multistatement_test.go:553]: ok

File: put_get_with_aws_test.go (hunk 16 above)
Comment (sfc-gh-pmotacki) [reply put_get_with_aws_test.go:212]: ok

File: rows_test.go (hunk 17 above)
Comment (sfc-gh-pmotacki) [reply rows_test.go:460]: ok

File: telemetry_test.go (hunk 20 above)
Comment (sfc-gh-pmotacki) [reply telemetry_test.go:137]: Better not. It is easier to read in separation

File: telemetry_test.go (hunk 22 above)
Comment (sfc-gh-pmotacki) [reply telemetry_test.go:246]: Better not. It is easier to read in separation

File: telemetry_test.go (hunk 21 above)
Comment (sfc-gh-pmotacki) [reply telemetry_test.go:207]: Better not. It is easier to read in separation

File: telemetry_test.go (hunk 19 above)
Comment (sfc-gh-pmotacki) [reply telemetry_test.go:107]: Better not. It is easier to read in separation

**Pull request merged**